File: src/haskell/4.hs
{-
[minhuang@mocha:haskell]$ time runghc problem-00004.hs
90909

real    0m0.280s
user    0m0.220s
sys     0m0.044s
-}
isPalindrome :: Integer -> Bool
isPalindrome x = (show x) == (reverse . show $ x)

result = head [x*y | x <- takeWhile (>0) [999,998..],
                     y <- takeWhile (>0) [x,x-1..],
                     isPalindrome (x*y)]
main = print result

File: ruby/4.rb
def is_palindrome(n)
  n.to_s == n.to_s.reverse
end

result = 0
999.step(1, -1).each do |i|
  i.step(1, -1).each do |j|
    x = i*j
    if is_palindrome(x) && x > result
      result = x
    end
  end
end
puts result
